### PR TITLE
Storage Topology CRD

### PR DESCRIPTION
**Pull request by dvaldivia**
Introduces `StorageTopology` CRD based on @wlan0 specification.

A lot of fields are optional right now, please let me know which ones are mandatory. Also the status fields are tentative.

```yaml
apiVersion: csi.min.io/v1alpha1
kind: StorageTopology
metadata:
  name: hdd
spec:
  layout: host{1...4}/hdd/data{1...4}
  fstype: xfs
  mount_options:
    - ro
    - atime
  resource_limit:
    storage: 10TiB
    volumes: 1000
  reclaim_policy: Retain|Delete
```

### CODE REVIEW

**Comment by wlan0** [comment]
I suggested using label selectors for nodes instead of hostnames to @abperiasamy, and we agreed that it was a superior approach as it can deal with node failure and automatically include new nodes meant for the old workload (expecting the same disk layout)

**Comment by wlan0** [comment]
This is how the new struct will look

```
apiVersion: direct.csi.min.io/v1
kind: StorageTopology
metadata:
  name: minio-drives
spec:
  layout:
    nodeLabels:
      - node.kubernetes.io/instance-type=i3en.24xlarge
      - node.kubernetes.io/region=us-east-1    
    path: /dev/nvme0n{1...8}
  fstype: xfs
  mount_options:
    - nouser
    - noatime
  resource_limit:
    storage: 8TiB
    volumes: 100
  reclaim_policy: Retain|Delete
```

**Comment by dvaldivia** [comment]
sure, I can update the CRD to contain that, what do you think about calling `nodeLabels` as `nodeSelector` since it would match the behavior?

**Comment by dvaldivia** [comment]
@wlan0 should `nodeLabels` be optional?

**Comment by dvaldivia** [comment]
@wlan0 done

**Comment by wlan0** [comment]
we use the api group `direct.csi.min.io`

https://github.com/minio/direct-csi/blob/master/pkg/volume/client.go#L46

File: pkg/apis/csi.min.io/v1alpha1/register.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+package v1alpha1
+
+import (
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// SchemeGroupVersion is group version used to register these objects
+var SchemeGroupVersion = schema.GroupVersion{Group: "cns.vmware.com", Version: "v1alpha1"}
```
Comment (wlan0) [inline pkg/apis/csi.min.io/v1alpha1/register.go:26]: The group cannot be `cns.vmware.com`

File: pkg/apis/csi.min.io/v1alpha1/types.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+type StorageTopologyStatusCapacity struct {
+	// Free Space of the storage topology
+	// +optional
+	FreeSpace resource.Quantity `json:"freeSpace,omitempty"`
+	// Total capacity of the storage topology
+	// +optional
+	Total resource.Quantity `json:"total,omitempty"`
+}
```
Comment (wlan0) [inline pkg/apis/csi.min.io/v1alpha1/types.go:113]: needs a field for the number of allocated volumes

Comment (wlan0) [reply pkg/apis/csi.min.io/v1alpha1/types.go:113]: call it `volumeAllocations`

Comment (dvaldivia) [reply pkg/apis/csi.min.io/v1alpha1/types.go:113]: just an integer?

Comment (wlan0) [reply pkg/apis/csi.min.io/v1alpha1/types.go:113]: It should go into `StorageTopologyStatus ` field

Comment (wlan0) [reply pkg/apis/csi.min.io/v1alpha1/types.go:113]: yeah, just an integer

Comment (dvaldivia) [reply pkg/apis/csi.min.io/v1alpha1/types.go]: done

**Comment by dvaldivia** [comment]
Current CR example:

```yaml
apiVersion: direct.csi.min.io/v1alpha1
kind: StorageTopology
metadata:
  name: minio-drives
spec:
  layout:
    nodeLabels:
      node.kubernetes.io/instance-type: i3en.24xlarge
      node.kubernetes.io/region: us-east-1
    path: /dev/nvme0n{1...8}
  fstype: xfs
  mount_options:
    - nouser
    - noatime
  resource_limit:
    storage: 8TiB
    volumes: 100
  reclaim_policy: Retain|Delete
```

File: pkg/apis/direct.csi.min.io/v1alpha1/register.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+package v1alpha1
+
+import (
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// SchemeGroupVersion is group version used to register these objects
+var SchemeGroupVersion = schema.GroupVersion{Group: "direc.csi.min.io", Version: "v1alpha1"}
```
Comment (wlan0) [inline pkg/apis/direct.csi.min.io/v1alpha1/register.go:26]: direct spelling

File: hack/boilerplate.go.txt [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// This file is part of MinIO Kubernetes Cloud
```
Comment (wlan0) [inline hack/boilerplate.go.txt:1]: Can you remove this change pls. This is unrelated. We can have this as a part of a separate PR if needed

Comment (dvaldivia) [reply hack/boilerplate.go.txt:1]: To regenerate the clients using `update-code-gen.sh` this file is needed as it provides the license for the generate files.

File: pkg/apis/direct.csi.min.io/v1alpha1/register.go (hunk 3 above)
Comment (dvaldivia) [reply pkg/apis/direct.csi.min.io/v1alpha1/register.go]: solved

File: hack/boilerplate.go.txt (hunk 4 above)
Comment (wlan0) [reply hack/boilerplate.go.txt:1]: got it

File: resources/direct.csi.min.io_storagetopologies.yaml [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
+          type: string
+        metadata:
+          description: 'Standard object metadata. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata'
+          type: object
+        spec:
+          description: 'Specification of the storage topology More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status.'
+          properties:
+            fstype:
+              description: Layout of the drives across hosts
```
Comment (wlan0) [inline resources/direct.csi.min.io_storagetopologies.yaml:36]: Filesystem type for formatting the underlying block devices

File: resources/direct.csi.min.io_storagetopologies.yaml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+              type: string
+            resource_limits:
+              description: Resource Limits
+              properties:
+                storage:
+                  anyOf:
+                  - type: integer
+                  - type: string
+                  description: Storage
+                  pattern: ^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$
```
Comment (wlan0) [inline resources/direct.csi.min.io_storagetopologies.yaml:68]: awesome!

**Comment by wlan0** [comment]
Only one minor issue with the text in the CRD definition. However, we can merge now and fix that later

thanks @dvaldivia !
